libraries/DataFlash/DataFlash_File.cpp

```diff
@@ -1091,36 +1051,34 @@ void DataFlash_File::_io_timer(void)
         // be kind to the FAT PX4 filesystem
         nbytes = _writebuf_chunk;
     }
-    if (_writebuf_head > _tail) {
-        // only write to the end of the buffer
-        nbytes = MIN(nbytes, _writebuf_size - _writebuf_head);
-    }
 
-    // try to align writes on a 512 byte boundary to avoid filesystem
-    // reads
+    uint32_t size;
+    const uint8_t *head = _writebuf.readptr(size);
+    nbytes = MIN(nbytes, size);
+
+    // try to align writes on a 512 byte boundary to avoid filesystem reads
     if ((nbytes + _write_offset) % 512 != 0) {
         uint32_t ofs = (nbytes + _write_offset) % 512;
         if (ofs < nbytes) {
             nbytes -= ofs;
         }
     }
 
-    assert(((uint32_t)_writebuf_head)+nbytes <= _writebuf_size);
-    ssize_t nwritten = ::write(_write_fd, &_writebuf[_writebuf_head], nbytes);
+    ssize_t nwritten = ::write(_write_fd, &head, nbytes);
     if (nwritten <= 0) {
         hal.util->perf_count(_perf_errors);
         close(_write_fd);
         _write_fd = -1;
-        _initialised = false;
+        _initialized = false;
     } else {
         _write_offset += nwritten;
+        _writebuf.advance(nwritten);
         /*
-          the best strategy for minimising corruption on microSD cards
+          the best strategy for minimizing corruption on microSD cards
```

Comment:
I prefer not to have a language war between english/US/australian spellings. Given its been done here then I won't ask it be changed, but please don't switch between different spellings as part of patch sets as it just clutters the patch